patches/tModLoader/Terraria/ModLoader/LocalizationLoader.cs

```diff
@@ -286,4 +295,317 @@ private static LocalizedText SetLocalizedText(Dictionary<string, LocalizedText>
+
+					CommentedWscJsonObject parent = rootObject;
+
+					string key = entry.key;
+					if (!string.IsNullOrWhiteSpace(localizationFileEntry.prefix)) {
+						key = key.Substring(localizationFileEntry.prefix.Length + 1);
+					}
+
+					// Find/Populate the parents of this translation entry 
+					string[] splitKey = key.Split(".");
+					for (int j = 0; j < splitKey.Length - 1; j++) {
+						string k = splitKey[j];
+						if (parent.ContainsKey(k))
+							parent = (CommentedWscJsonObject)parent[k];
+						else {
+							var newParent = new CommentedWscJsonObject();
+							parent.Add(k, newParent);
+							parent = newParent;
+						}
+					}
+
+					// TODO: "$parentVal" support?
+					// Populate parent object with this translation, manipulating comments to appear above the entry.
+
+					if (entry.value == null && entry.type == JsonType.Object) {
+						if (parent.Count == 0) {
+							parent.Comments[""] = entry.comment;
+						}
+						else {
+							string actualCommentKey = parent.Keys.Last();
+							parent.Comments[actualCommentKey] = entry.comment;
+						}
+						parent.Add(key.Split(".").Last(), new CommentedWscJsonObject());
+					}
+					else {
+						// Add values
+
+						string value = translations[entry.key].GetTranslation(culture);
+						key = key.Split(".").Last();
+						if (culture.Name != "en-US" && value == translations[entry.key].GetDefault()) {
+							// This might be messing up Russian: OctopusBanner: "{$CommonItemTooltip.BannerBonus}{$Mods.ExampleMod.NPCName.Octopus}"
+							//key = "# " + key; // doesn't work, escaped by escapeName
+							parent.CommentedOut.Add(key);
+						}
+
+						if (parent.Count == 0) {
+							parent.Comments[""] = entry.comment;
+						}
+						else {
+							string actualCommentKey = parent.Keys.Last();
+							parent.Comments[actualCommentKey] = entry.comment;
+						}
+						parent.Add(key, value);
+					}
+				}
+
+				string outputFileName = localizationFileEntry.path;
+				outputFileName = outputFileName.Replace("en-US", culture.CultureInfo.Name);
+				string outputFilePath = Path.Combine(sourceFolder, outputFileName);
+				outputFilePath += ".new"; // Save to new file, until working completely
+
+				string hjsonContents = rootObject.ToFancyHjsonString();
+				Directory.CreateDirectory(Path.GetDirectoryName(outputFilePath)); // Folder might not exist when using Extract mode
+				File.WriteAllText(outputFilePath, hjsonContents);
+
+				allLocalizationFilesAllLanguages.Remove(outputFileName);
+
+				// TODO: Indicate on Mods/Mod Sources that localizations have updated maybe?
+			}
+		}
+
+		if (specificCulture != null)
+			return;
+
+		// Clean up orphaned non-default language files, if any.
+		foreach (var remainingLocalizationFiles in allLocalizationFilesAllLanguages) {
+			string originalPath = Path.Combine(sourceFolder, remainingLocalizationFiles);
+			string newPath = originalPath + ".legacy";
+
+			File.Move(originalPath, newPath);
+		}
```

Comment:
Hmm, this gets a bit tricker with the "don't update files which didn't change" feature, but I have an idea.

This only deletes files which don't have an English counterpart, so you could attempt to map back the other way (find an english file with the appropriate name), or map all the english files to each culture to generate the full set of known paths.

Reply:
This does that, the only files left in here will be those that didn't match up to an english file.